### PR TITLE
Add Intrinsic Curiosity Module

### LINKED ISSUE #502: Intrinsic Curiosity Module (ICM)
- [x] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [x] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```

I'd like to contribute an implementation of Intrinsic Curiosity Module (ICM). ([paper](https://arxiv.org/pdf/1705.05363.pdf), [author's repo](https://github.com/pathak22/noreward-rl/), [reference repo](https://github.com/uvipen/Street-fighter-A3C-ICM-pytorch).)

### PR DESCRIPTION
**Pull request by nuance1979**
- [x] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [ ] algorithm implementation fix
    + [ ] documentation modification
    + [x] new feature
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the code using `make commit-checks` (**required**)
- [x] If applicable, I have mentioned the relevant/related issue(s)
- [x] If applicable, I have listed every items in this Pull Request below

Implementation of Intrinsic Curiosity Module (ICM, #502)

Paper: https://arxiv.org/pdf/1705.05363.pdf
Reference repos: 
 - https://github.com/pathak22/noreward-rl/
 - https://github.com/uvipen/Street-fighter-A3C-ICM-pytorch

ICM is implemented as a decorator because it can be used with any underlying policy. I have prepared an example with atari dqn for simplicity but it might not be able to give better performance for many Atari envs. I'm trying to get an example with VizDoom env running.

### CODE REVIEW

File: tianshou/policy/modelbased/icm.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            batch == Batch(
+                obs=Batch(
+                    obs="original obs, with batch_size=1 for demonstration",
+                    mask=np.array([[False, True, False]]),
+                    # action 1 is available
+                    # action 0 and 2 are unavailable
+                ),
+                ...
+            )
+
```
Comment (Trinkle23897) [inline tianshou/policy/modelbased/icm.py]:
```suggestion
```

File: tianshou/policy/modelbased/icm.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        :param float eps: in [0, 1], for epsilon-greedy exploration method.
+
+        :return: A :class:`~tianshou.data.Batch` which has 3 keys:
+
+            * ``act`` the action.
+            * ``logits`` the network's raw output.
+            * ``state`` the hidden state.
```
Comment (Trinkle23897) [inline tianshou/policy/modelbased/icm.py]: change this

File: tianshou/policy/modelbased/icm.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        inverse_loss = F.cross_entropy(act_hat, act).mean()
+        forward_loss = batch.policy.mse_loss.mean()
+        loss = (
+            (1 - self.forward_loss_weight) * inverse_loss +
+            self.forward_loss_weight * forward_loss
+        ) * self.lr_scale
+        loss.backward()
+        self.optim.step()
+        return {
+            "loss/policy": res["loss"],
```
Comment (Trinkle23897) [inline tianshou/policy/modelbased/icm.py]: This could cause problems with other policies that don't have `loss`. Maybe switch to 
```python
res.update({"loss/icm": ..., "loss/icm/forward": ..., ...})
return res
```

File: examples/vizdoom/vizdoom_a2c_icm.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        torch.save(policy.state_dict(), os.path.join(log_path, 'policy.pth'))
+
+    def stop_fn(mean_rewards):
+        if env.spec.reward_threshold:
+            return mean_rewards >= env.spec.reward_threshold
+        elif 'Pong' in args.task:
+            return mean_rewards >= 20
+        else:
+            return False
+
```
Comment (Trinkle23897) [inline examples/vizdoom/vizdoom_a2c_icm.py]:
```suggestion
        return False
```

File: tianshou/policy/modelbased/icm.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data.
+
+        :return: A :class:`~tianshou.data.Batch` which has at least 3 keys:
+
+            * ``act`` the action.
+            * ``logits`` the network's raw output.
+            * ``state`` the hidden state.
```
Comment (Trinkle23897) [inline tianshou/policy/modelbased/icm.py]:
```suggestion
        """Compute action over the given batch data by inner policy.
```

**Review by Trinkle23897 — APPROVED** [review]
